RabbitMQ Secret Backend

KevinPike opened the pull request:
This adds a RabbitMQ secret backend modeled after the Postgresql secret backend. RabbitMQ connections consist of a URI, username, and password. Roles contain tags and vhost permissions.

jefferai commented:
Hi @KevinPike ,

Thanks for submitting this. I'm going to mark this as "future" -- it is past the 0.4 merge window and we are planning for very large changes to the APIs for backends starting with 0.5. Once that's finalized, hopefully you could port this to use the new API and we could get this merged.

Thanks!

KevinPike commented:
Hey @jefferai,

Thanks for the update! Sounds good to me, I'll revisit this when the 0.5 backend APIs are out.

jefferai commented:
Hi @KevinPike ,

We'd like to get this into 0.5. Any chance you can work on documentation for it? I'll be doing code review today or tomorrow.

Thanks!

KevinPike commented:
Hi @jefferai, I just saw Vault was reached 0.5. I should have time to work on documentation over the next couple of days.

jefferai commented:
Neato! We plan on having 0.5.1 pretty soon and 0.5.2 coming a few weeks after that.

KevinPike commented:
I added the documentation, updated to master, and updated changes in the backend. What's the best way to go about adding `github.com/michaelklishin/rabbit-hole` to `vendor/`? Do I just copy the repository into vendor?

vishalnayak commented:
@KevinPike 
I checked out your PR and following are the steps that worked for me.
1) Make sure you have Go1.6 installed.
2) Have the latest version of godep
`go get -u github.com/tools/godep`
3) Remove the Godeps workspace.
`rm -rf Godeps/_workspace`
3) Save the dependencies in vendor directory
`godep save ./...`
4) Step 3 would remove the existing dependencies in the vendor folder. Revert it.
`git checkout .`
5) Add the new vendor dependencies
`git add .`

I cannot assure you that the same steps will work for you, but it's worth a try!
Let me know.

vishalnayak commented on builtin/logical/rabbitmq/backend.go:
```diff
@@ -0,0 +1,118 @@
+		Paths: []*framework.Path{
+			pathConfigConnection(&b),
+			pathConfigLease(&b),
+			pathRoles(&b),
```
Can you consider adding `pathListRoles` as well, just like `postgresql` backend?

vishalnayak commented on builtin/logical/rabbitmq/backend.go:
```diff
@@ -0,0 +1,118 @@
+func (b *backend) ResetClient() {
+	b.lock.Lock()
+	defer b.lock.Unlock()
+
```
Just to be sure, is there no need to close any connections, delete users/policies/vhosts here?

KevinPike replied:
@vishalnayak I checked the methods on `*rabbithole.Client` and didn't find a `Close` method so I don't think any connections need to be closed. I think this client is talking to the RabbitMQ admin API. 

As far as deleting users/policies/vhosts, is there anything you suggest? I checked the PostgreSQL backend in hopes to see how it deletes things and I see it's not deleting anything.

vishalnayak replied:
@KevinPike Regarding `rabbithole.Client`'s `Close` method, I saw this: https://godoc.org/github.com/michaelklishin/rabbit-hole#Client.CloseConnection. Maybe it is worth checking out!

Regarding deleting users/policies/vhosts, I only wanted your second look at it, just in case it was missed out. Usually, it should be taken care of in the secrets revocation scenario, rather than here, which I see that I have commented on already anyways. So, I guess this is good enough.

Just check the `Close` on client.

KevinPike replied:
Thanks @vishalnayak. `rabbithole.Client`'s `CloseConnection` closes connection between a RabbitMQ server and RabbitMQ client. Here's an [example](https://github.com/michaelklishin/rabbit-hole#operations-on-connections). I don't think it applies here.

jefferai replied:
I think we should explicitly close connections whenever we are changing the state of the client (before setting to nil, or before setting to a new client). Go has a bad habit of leaking file descriptors when network connections that aren't closed properly. This should happen in the Cleanup func too.

Here's a link to the postgres revocation code: https://github.com/hashicorp/vault/blob/master/builtin/logical/postgresql/secret_creds.go#L85

jefferai replied:
Took another look. I do think `CloseConnection` needs to be called when resetting or cleaning up the backend. It tells the client to close its connection to the server, not the other way around.

vishalnayak commented on builtin/logical/rabbitmq/path_config_connection.go:
```diff
@@ -0,0 +1,109 @@
+	return &framework.Path{
+		Pattern: "config/connection",
+		Fields: map[string]*framework.FieldSchema{
+			"uri": &framework.FieldSchema{
```
Can we rename this to `connection_uri`?

vishalnayak commented on builtin/logical/rabbitmq/path_config_connection.go:
```diff
@@ -0,0 +1,109 @@
+			"password": &framework.FieldSchema{
+				Type:        framework.TypeString,
+				Description: "Password of the provided RabbitMQ management user",
+			},
```
There is a need for `verify_connection` option (refer postgresql backend).
This option makes the `connection_uri` verification below, toggle-able, during the configuration.

KevinPike replied:
Updated, thanks!

vishalnayak commented on builtin/logical/rabbitmq/path_config_connection.go:
```diff
@@ -0,0 +1,109 @@
+
+const pathConfigConnectionHelpDesc = `
+This path configures the connection properties used to connect to RabbitMQ management HTTP API.
+The "uri" parameter is a string that is be used to connect to the API. The "username"
```
"that is be used" --> "that is used"

vishalnayak commented on builtin/logical/rabbitmq/path_config_connection.go:
```diff
@@ -0,0 +1,109 @@
+const pathConfigConnectionHelpDesc = `
+This path configures the connection properties used to connect to RabbitMQ management HTTP API.
+The "uri" parameter is a string that is be used to connect to the API. The "username"
+and "password" parameters are strings and used as credentials to the API.
```
"strings and used" --> "strings that are used"

vishalnayak commented on builtin/logical/rabbitmq/path_config_connection.go:
```diff
@@ -0,0 +1,109 @@
+The URI looks like:
+"http://localhost:15672"
+
+When configuring the URI, username, and password, the backend will verify their validity.
```
We can mention that this is toggle-able. i.e. after the addition of `verify_connection` option.

vishalnayak commented on builtin/logical/rabbitmq/path_role_create.go:
```diff
@@ -0,0 +1,119 @@
+		lease = &configLease{Lease: 1 * time.Hour}
+	}
+
+	// Generate the username, password and expiration. PG limits user to 63 characters
```
We can update this comment.

vishalnayak commented on builtin/logical/rabbitmq/secret_creds.go:
```diff
@@ -0,0 +1,74 @@
+		return nil, err
+	}
+
+	_, err = client.DeleteUser(username)
```
If the user is deleted, will all the associated vhosts get deleted?

KevinPike replied:
This backend does not create vhosts if they do not exist. Users are given permission to the vhosts provided.

jefferai replied:
Sounds good -- leave the vhosts alone.

vishalnayak commented on website/source/docs/secrets/rabbitmq/index.html.md:
```diff
@@ -0,0 +1,326 @@
+---
+layout: "docs"
+page_title: "Secret Backend: RabbitMQ"
+sidebar_current: "docs-secrets-rabbitmq"
```
There should be an entry in `website/source/layouts/docs.erb` for `docs-secrets-rabbitmq` for this to work.

vishalnayak replied:
In fact you can try running the local website server and actually test this to avoid surprises.

Have `Ruby 2.x`, and `bundler` installed.
Then run the following:
cd website
`bundle`
`bundle exec middleman server`

vishalnayak commented on website/source/docs/secrets/rabbitmq/index.html.md:
```diff
@@ -0,0 +1,326 @@
+questionable data access is discovered: you can track it down to the specific
+instance of a service based on the RabbitMQ username.
+
+// TODO: Fix this
```
Just a reminder that there is a TODO here :)

vishalnayak commented on website/source/docs/secrets/rabbitmq/index.html.md:
```diff
@@ -0,0 +1,326 @@
+
+## Quick Start
+
+The first step to using the PostgreSQL backend is to mount it.
```
"PostgreSQL" --> "RabbitMQ"

vishalnayak commented on website/source/docs/secrets/rabbitmq/index.html.md:
```diff
@@ -0,0 +1,326 @@
+  <dd>POST</dd>
+
+  <dt>URL</dt>
+  <dd>`/RabbitMQ/config/connection`</dd>
```
Lower case "rabbitmq"

vishalnayak commented on website/source/docs/secrets/rabbitmq/index.html.md:
```diff
@@ -0,0 +1,326 @@
+      <li>
+        <span class="param">uri</span>
+        <span class="param-flags">required</span>
+        The RabbitMQ management connection URI. e.g. "http://localhost:15672"
```
Since the example is mentioned in the texts above, there might not be a need to mention it in the API section.

vishalnayak commented on website/source/docs/secrets/rabbitmq/index.html.md:
```diff
@@ -0,0 +1,326 @@
+      <li>
+        <span class="param">username</span>
+        <span class="param-flags">required</span>
+        The RabbitMQ management administrator username. e.g. "admin"
```
Same here. Example not needed.

vishalnayak commented on website/source/docs/secrets/rabbitmq/index.html.md:
```diff
@@ -0,0 +1,326 @@
+      <li>
+        <span class="param">password</span>
+        <span class="param-flags">required</span>
+        The RabbitMQ management administrator password. e.g. "password"
```
Same here. Example not needed.

jefferai commented:
Hi @KevinPike,

Feel free to just drop the Godeps changes altogether; I can handle the vendoring once it's ready to go.

KevinPike commented:
Hi @vishalnayak and @jefferai, I believe I have addresses all of your comments. Thank you for thorough review!

jefferai commented on builtin/logical/rabbitmq/backend.go:
```diff
@@ -0,0 +1,119 @@
+		Help: strings.TrimSpace(backendHelp),
+
+		PathsSpecial: &logical.Paths{
+			Root: []string{
```
We've generally moved away from root paths as the ACL system is very flexible and allows `sudo` to be given on top of other sets of capabilities. So you can remove this and just let the ACL system do its job. :-)

jefferai commented on builtin/logical/rabbitmq/path_config_connection.go:
```diff
@@ -0,0 +1,120 @@
+		if err != nil {
+			return logical.ErrorResponse(fmt.Sprintf(
+				"Error validating connection info by listing users: %s", err)), nil
+		}
```
The connection should be closed here too.

KevinPike replied:
The only close method on the client is `CloseConnection` and it closes other connections, not it's own. Here's an example from the godoc:
```go
// Forcefully close connection
_, err := rmqc.CloseConnection("127.0.0.1:50545 -> 127.0.0.1:5672")
```

jefferai replied:
A client can have more than one connection associated with it. I guess my thinking is, if the client has changed, *all* connections for that client should be closed, and the client re-opened. Or in the case of verifying, the connections should all be closed and the client torn down.

What we want to avoid is connections not closing properly and causing leaked file descriptors. This has hit us with a number of other backends due to connections not being closed properly.

jefferai commented on builtin/logical/rabbitmq/path_config_lease.go:
```diff
@@ -0,0 +1,103 @@
+	return &framework.Path{
+		Pattern: "config/lease",
+		Fields: map[string]*framework.FieldSchema{
+			"lease": &framework.FieldSchema{
```
Please use `ttl` and `ttl_max` -- since about 0.3 we've been using these to avoid confusion with the concept of the `lease` itself.

jefferai commented on builtin/logical/rabbitmq/path_config_lease.go:
```diff
@@ -0,0 +1,103 @@
+		Pattern: "config/lease",
+		Fields: map[string]*framework.FieldSchema{
+			"lease": &framework.FieldSchema{
+				Type:        framework.TypeString,
```
Please use `framework.TypeDurationSecond` as it accepts both strings and integers and provides automatic conversion.

jefferai commented on builtin/logical/rabbitmq/path_config_lease.go:
```diff
@@ -0,0 +1,103 @@
+
+func (b *backend) pathLeaseWrite(
+	req *logical.Request, d *framework.FieldData) (*logical.Response, error) {
+	leaseRaw := d.Get("lease").(string)
```
With `framework.TypeDurationSecond` you'll want to use `.(int)` here, which will give a value in seconds.

jefferai commented on builtin/logical/rabbitmq/path_config_lease.go:
```diff
@@ -0,0 +1,103 @@
+	leaseRaw := d.Get("lease").(string)
+	leaseMaxRaw := d.Get("lease_max").(string)
+
+	lease, err := time.ParseDuration(leaseRaw)
```
Then you can avoid the ParseDuration calls and simply (if one or the other is actually set and not 0) set it to `time.Second * <value>`

jefferai commented on builtin/logical/rabbitmq/path_roles.go:
```diff
@@ -0,0 +1,173 @@
+
+func (b *backend) pathRoleDelete(
+	req *logical.Request, data *framework.FieldData) (*logical.Response, error) {
+	err := req.Storage.Delete("role/" + data.Get("name").(string))
```
You should check here to make sure that the role name is not empty, otherwise you risk doing a prefix delete on `role/` which could be very bad.

jefferai commented on builtin/logical/rabbitmq/path_roles.go:
```diff
@@ -0,0 +1,173 @@
+
+func (b *backend) pathRoleCreate(
+	req *logical.Request, data *framework.FieldData) (*logical.Response, error) {
+	name := data.Get("name").(string)
```
You should make sure the name isn't empty.

jefferai commented on builtin/logical/rabbitmq/path_roles.go:
```diff
@@ -0,0 +1,173 @@
+func (b *backend) pathRoleCreate(
+	req *logical.Request, data *framework.FieldData) (*logical.Response, error) {
+	name := data.Get("name").(string)
+	tags := data.Get("tags").(string)
```
Any reason not to split the tags here into an array of strings, and store that instead?

jefferai commented on builtin/logical/rabbitmq/secret_creds.go:
```diff
@@ -0,0 +1,75 @@
+		return nil, err
+	}
+	if lease == nil {
+		lease = &configLease{Lease: 1 * time.Hour}
```
Don't set this to a value -- set it to `lease = &configLease{}`. The reason is that that way, if people don't set a default TTL in the backend but do set a default TTL on the mount, it will be honored.

jefferai commented on builtin/logical/rabbitmq/secret_creds.go:
```diff
@@ -0,0 +1,75 @@
+	if !ok {
+		return nil, fmt.Errorf("secret is missing username internal data")
+	}
+	username, ok := usernameRaw.(string)
```
You're not actually checking this `ok` value here. I think it's fine to ignore the check at this point and just type assert it directly, but if you're going to use the `ok` form, you should check the `ok` value.

jefferai commented on builtin/logical/rabbitmq/path_role_create.go:
```diff
@@ -0,0 +1,111 @@
+
+func (b *backend) pathRoleCreateRead(
+	req *logical.Request, data *framework.FieldData) (*logical.Response, error) {
+	name := data.Get("name").(string)
```
You should check that `name` isn't empty.

jefferai commented on builtin/logical/rabbitmq/path_role_create.go:
```diff
@@ -0,0 +1,111 @@
+
+	displayName := req.DisplayName
+	if len(displayName) > 26 {
+		displayName = displayName[:26]
```
You should create a `logical.Response` object above and, if this happens, use `AddWarning()` on the response object to let the user know that it was truncated.

jefferai commented on builtin/logical/rabbitmq/path_role_create.go:
```diff
@@ -0,0 +1,111 @@
+	client, err := b.Client(req.Storage)
+	if err != nil {
+		return nil, err
+	}
```
I realize that `b.Client` won't return a nil client with your current code, but for future safety's sake, it'd be good to ensure that `client != nil` here.

jefferai commented on builtin/logical/rabbitmq/secret_creds.go:
```diff
@@ -0,0 +1,75 @@
+	username, ok := usernameRaw.(string)
+
+	// Get our connection
+	client, err := b.Client(req.Storage)
```
I realize that `b.Client` won't return a nil client with your current code, but for future safety's sake, it'd be good to ensure that `client != nil` here.

jefferai commented on builtin/logical/rabbitmq/path_role_create.go:
```diff
@@ -0,0 +1,111 @@
+	}
+
+	// Return the secret
+	resp := b.Secret(SecretCredsType).Response(map[string]interface{}{
```
It'd be good to add the username into `Metadata`. This gets logged in plaintext in audit logs so can help someone trying to do forensics or audit searching later on.

KevinPike replied:
I'm not sure how add username into `Metadata`. username is in the second parameter to `Response()`, is that the same?

jefferai replied:
Metadata is a field in the response. So you should be able to add it into `resp.Metadata["username"]`.

jefferai commented on builtin/logical/rabbitmq/path_role_create.go:
```diff
@@ -0,0 +1,111 @@
+			Read:      permission.Read,
+		})
+
+		if err != nil {
```
I assume that if this happens, we want to actually attempt to remove the user, or the user will be created and sitting in RMQ forever.

jefferai commented:
Hi @KevinPike ,

I'm really, really sorry it's taken so long for me to get to this. This is looking pretty solid -- I left a bunch of comments but they're mostly very minor things.

0.5.3 should land sometime next week and hopefully this can be a part of it!

jippi commented:
👍  !

adamlc commented:
Looking forward to this 💃

KevinPike commented:
Hi @jefferai,

It's all good, it takes me awhile too. Thank you for another thorough review :smile: 

As a general question, when is it good to use `logical.ErrorResponse` and `logical.HelpResponse` in backend `logical.Response`s? At first, I thought it was good to use when you don't have an explicit error, like `role == nil`. Now, I'm thinking it's better use to than not in error cases because more context can be added to `logical.ErrorResponse`.

jefferai commented:
Hi @KevinPike,

HelpResponses are generated automatically when using path-help. The problem with using them in error cases is it will cause a 200 instead of an error status code. For any Response you can use AddWarning, which will keep the appropriate status code.

jefferai commented:
Hi @KevinPike ,

To make it easier to see what's left, I deleted some of the older feedback that hadn't been closed automatically by code changes.

It seems the only two items are around connection closing and putting the username in metadata. The latter should be pretty trivial; the former, I think the issue is I don't really see why you think it's a bad idea to close the client's connections (even if it's all of them) when we're modifying the client (or deleting it altogether, for instance, after a connection check). In the latter case we're not reusing the client, and in the former case it seems like we should simply close the connections on the original client and recreate a new client with the new parameters. Am I missing something?

jefferai commented:
Hi @KevinPike,

It'd be stellar if we can land this in the next day or two. Any chance you can look at the comments before then?

vishalnayak commented:
@KevinPike 

Many thanks for submitting this, dealing with this for long time and for working on all the review comments. I am going to merge this!
